### PR TITLE
Flask: Groups endpoints

### LINKED ISSUE #192: Epic: /api/groups
Create a new file `groups.py` for these endpoints

- [x] **GET /api/groups** (list_groups)
  Requires authentication. Returns a list of all group codes and display names, e.g. `[{ "group_code": "ACH", "group_name": "Alberta" }]`

- [x] **GET /api/groups/:group_code** (get_group)
  Returns `{ "group_code": "", "group_name": "",  "users": [] }` where each field is filled out accordingly. The `users` array should be an array of usernames of the users in the group and can be obtained from the `users` relation on the Group model as a backref from User. Admins can use this endpoint for all groups. Regulars can only use this for groups they belong to or they get a 403 Forbidden.

- [x] **POST /api/groups** (create_group)
  Admin-only. Creates a new group with the specified codename, display name, and users (same format as the get_group endpoint return result, but the users array is optional). This also creates a corresponding MinIO group (via madmin.MinioAdmin) and bucket (via MinioClient). The group should have an access policy that gives it complete access to its bucket. The corresponding MinIO users should be added to this group. Fail with 422 if the bucket already exists (do not create a group in the database in this case) or the code/name are already in use. Fail with 404 if a user to be added doesn't exist. The return result should be of the same form as `get_group` and the URL to the created resource should be in the Location header.

- [x] **PATCH /api/groups** (update_groups)
  Admin-only. Updates requested fields accordingly. The group code cannot be changed but the display name can be. Users should be added and removed accordingly and the error codes are the same as create_group.

- [x] **DELETE /api/groups/:group_code** (delete_group)
  Admin-only. Deletes the group and the MinIO group if they are empty but retains the MinIO bucket, otherwise fail with 422.

### PR DESCRIPTION
**Pull request by SamarthVP**
closes #192 

Sample tests, will not work as expected unless User.minio_access_key and User.minio_secret_key are manually set to a manually created minio user:
```
curl -D - -X POST \
        -H "Content-Type: application/json" \
        --data '{"group_name" : "bruh20", "group_code" : "whoknows20",
                "users": [
                "admin"
                ]
                }' \
        localhost:5000/api/groups

curl -D - -X PATCH \
        -H "Content-Type: application/json" \
        --data '{"group_name" : "changed20.1",
                "users": [
                
                ]
                }' \
        localhost:5000/api/groups/whoknows20

curl -D - -X DELETE \
        localhost:5000/api/groups/whoknows20
```

### CODE REVIEW

File: flask/app/groups.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        if request.json["group_name"]:
+            # Check if display name is in use, 422
+            conflicting_group = models.Group.query.filter_by(
+                group_name=request.json["group_name"]
+            ).one_or_none()
+            if (conflicting_group is not None) and (conflicting_group != group):
+                return "Group name in use", 422
+            group.group_name = request.json["group_name"]
+
+    minioAdmin = MinioAdmin(
```
Comment (kevinlul) [inline flask/app/groups.py:89]: Please use snake_case since this is in Python

File: flask/app/groups.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        ).all()
+        # Make sure all users are valid
+        if len(users) != len(request.json["users"]):
+            return "Invalid username provided", 404
+
+        # Clear users from db and minio group
+        if len(group.users) != 0:
+            group_info = minioAdmin.get_group(group_code)
+            for old_user in group_info["members"]:
+                minioAdmin.group_remove(group_code, old_user)
```
Comment (kevinlul) [inline flask/app/groups.py:108]: This function supports passing the entire list of users in at once if you use the spread operator, i.e. `*group_info["members"]`

File: flask/app/groups.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            group.users[:] = []
```
Comment (kevinlul) [inline flask/app/groups.py:109]: What's the use of the slice operator here?

File: flask/app/groups.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            minioAdmin.group_add(group_code, user.minio_access_key)
+        # Reset policy for group in case the group did not exist
+        minioAdmin.set_policy(group_code, group=group_code)
+
+    try:
+        db.session.commit()
+        return jsonify(group)
+    except:
+        db.session.rollback()
+        return "Server error", 500
```
Comment (kevinlul) [inline flask/app/groups.py:124]: Hope this never happens because we can't roll back the MinIO side in the same way 😓

File: flask/app/groups.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+
+@app.route("/api/groups", methods=["POST"])
+@login_required
+@check_admin
+def create_group():
+    if not request.json:
+        return "Request body must be JSON", 400
```
Comment (kevinlul) [inline flask/app/groups.py:132]: Should be 415

File: flask/app/groups.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+    if not group_name:
+        return "A group display name must be provided", 400
+    if not group_code:
+        return "A group codename must be provided", 400
+
+    if models.Group.query.filter(
+        (models.Group.group_code == group_code)
+        | (models.Group.group_name == group_name)
+    ).value("group_id"):
```
Comment (kevinlul) [inline flask/app/groups.py:145]: This is interesting that you can use the pipe operator like this!

File: flask/app/groups.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    # Add users to the db if they were given
+    if request.json.get("users"):
+        users = models.User.query.filter(
+            models.User.username.in_(request.json["users"])
+        ).all()
+        # Make sure all users are valid
+        if len(users) != len(request.json["users"]):
+            return "Invalid username provided", 404
+        for user in users:
+            group_obj.users.append(user)
```
Comment (kevinlul) [inline flask/app/groups.py:159]: as an optional thing, could you check if Python list concatenation works with SQLAlchemy collections?

File: flask/app/groups.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        endpoint=app.config["MINIO_ENDPOINT"],
+        access_key=app.config["MINIO_ACCESS_KEY"],
+        secret_key=app.config["MINIO_SECRET_KEY"],
+    )
+
+    # Create minio bucket via mc, 422 if it already exists
+    if minioClient.bucket_exists(group_code):
+        return "Minio bucket already exists", 422
+
+    minioClient.make_bucket(group_code)
```
Comment (kevinlul) [inline flask/app/groups.py:178]: What happens when the `group_code` contains invalid characters or is too short to be a bucket name? Does the SDK raise an exception?

**Review by kevinlul — CHANGES_REQUESTED** [review]

File: flask/app/groups.py (hunk 1 above)
Comment (SamarthVP) [reply flask/app/groups.py]: Old habits lol

File: flask/app/groups.py (hunk 2 above)
Comment (SamarthVP) [reply flask/app/groups.py]: :O i didn't know this was a thing

File: flask/app/groups.py (hunk 3 above)
Comment (SamarthVP) [reply flask/app/groups.py]: It's unnecessary, good catch

File: flask/app/groups.py (hunk 5 above)
Comment (SamarthVP) [reply flask/app/groups.py]: Fixed

File: flask/app/groups.py (hunk 7 above)
Comment (SamarthVP) [reply flask/app/groups.py]: Wow it actually does, that's much cleaner

File: flask/app/groups.py (hunk 6 above)
Comment (SamarthVP) [reply flask/app/groups.py:144]: I believe filter() just takes any conditional statement so | is treated as the or operator

File: flask/app/groups.py (hunk 8 above)
Comment (SamarthVP) [reply flask/app/groups.py]: True, both `bucket_exists()` and `make_bucket()` would throw errors which we should definitely be checking for. I've moved those commands closer to the beginning and if they error I return "Invalid bucket name", 422. I've also added a test case to test for this.

File: flask/tests/test_groups.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
+import pytest
+from app import db, models
+from sqlalchemy.orm import joinedload
+from conftest import TestConfig
+from app.madmin import MinioAdmin, readwrite_buckets_policy
+from minio import Minio
+
+
+@pytest.fixture
+def minioAdmin():
```
Comment (kevinlul) [inline flask/tests/test_groups.py:10]: snake_case

File: flask/tests/test_groups.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+    )
+    # Test group_code already in use
+    assert (
+        client.post(
+            "/api/groups",
+            json={"group_code": "ach", "group_name": "Nothing suspicious"},
+        ).status_code
+        == 422
+    )
+    # Test invlaid bucket name
```
Comment (kevinlul) [inline flask/tests/test_groups.py:203]: typo

File: flask/tests/test_groups.py [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+    # Test existence
+    assert (
+        client.patch("/api/groups/code", json={"group_name": "yes"}).status_code == 404
+    )
+
+    # Test success and check db and minio
+    ad = models.User.query.filter(models.User.username == "admin").one_or_none()
+    ad.minio_access_key = "admin"
+    user = models.User.query.filter(models.User.username == "user").one_or_none()
+    user.minio_access_key = "user"
```
Comment (kevinlul) [inline flask/tests/test_groups.py:155]: This might be better for the test_database fixture instead of making use of the shared SQLAlchemy session here.

**Review by kevinlul — CHANGES_REQUESTED** [review]
Overall, this looks really good!

File: flask/tests/test_groups.py (hunk 11 above)
Comment (SamarthVP) [reply flask/tests/test_groups.py]: Good point, i thought it might break other tests but it should not and does not.

**Review by kevinlul — APPROVED** [review]
Solid!

**Pull request merged**